src/mito2/src/memtable/bulk/part.rs

```diff
@@ -555,29 +593,113 @@ pub struct BulkPartMeta {
     pub region_metadata: RegionMetadataRef,
 }
 
+/// Metrics for encoding a part.
+#[derive(Default, Debug)]
+pub struct BulkPartEncodeMetrics {
+    /// Cost of iterating over the data.
+    pub iter_cost: Duration,
+    /// Cost of writing the data.
+    pub write_cost: Duration,
+    /// Size of data before encoding.
+    pub raw_size: usize,
+    /// Size of data after encoding.
+    pub encoded_size: usize,
+    /// Number of rows in part.
+    pub num_rows: usize,
+}
+
 pub struct BulkPartEncoder {
     metadata: RegionMetadataRef,
     row_group_size: usize,
     writer_props: Option<WriterProperties>,
 }
 
 impl BulkPartEncoder {
-    pub(crate) fn new(metadata: RegionMetadataRef, row_group_size: usize) -> BulkPartEncoder {
+    pub(crate) fn new(
+        metadata: RegionMetadataRef,
+        row_group_size: usize,
+    ) -> Result<BulkPartEncoder> {
+        // TODO(yingwen): Skip arrow schema if needed.
+        let json = metadata.to_json().context(InvalidMetadataSnafu)?;
+        let key_value_meta =
+            parquet::file::metadata::KeyValue::new(PARQUET_METADATA_KEY.to_string(), json);
+
+        // TODO(yingwen): Do we need compression?
         let writer_props = Some(
             WriterProperties::builder()
+                .set_key_value_metadata(Some(vec![key_value_meta]))
                 .set_write_batch_size(row_group_size)
                 .set_max_row_group_size(row_group_size)
                 .build(),
         );
-        Self {
+
+        Ok(Self {
             metadata,
             row_group_size,
             writer_props,
-        }
+        })
     }
 }
 
 impl BulkPartEncoder {
+    /// Encodes [BoxedRecordBatchIterator] into [EncodedBulkPart] with min/max timestamps.
+    pub fn encode_record_batch_iter(
+        &self,
+        iter: BoxedRecordBatchIterator,
+        arrow_schema: SchemaRef,
+        min_timestamp: i64,
+        max_timestamp: i64,
+        metrics: &mut BulkPartEncodeMetrics,
+    ) -> Result<Option<EncodedBulkPart>> {
+        let mut buf = Vec::with_capacity(4096);
+        let mut writer = ArrowWriter::try_new(&mut buf, arrow_schema, self.writer_props.clone())
+            .context(EncodeMemtableSnafu)?;
+        let mut total_rows = 0;
+
+        // Process each batch from the iterator
+        let mut iter_start = Instant::now();
+        for batch_result in iter {
+            metrics.iter_cost += iter_start.elapsed();
+            let batch = batch_result?;
+            if batch.num_rows() == 0 {
+                continue;
+            }
+
+            metrics.raw_size += batch.get_array_memory_size();
```

Comment:
`batch.get_array_memory_size()` can be far away from accurate, see also https://github.com/GreptimeTeam/greptimedb/pull/6550 if we do need that metric.